clippy_lints/src/redundant_local.rs

```diff
@@ -0,0 +1,136 @@
+use clippy_utils::{diagnostics::span_lint_and_help, source::snippet_opt};
+use rustc_ast::{
+    ast::{BindingAnnotation, Block, Expr, ExprKind, LocalKind, Mutability, Pat, PatKind, Path, StmtKind},
+    visit::FnKind,
+    Local,
+};
+use rustc_data_structures::fx::FxHashMap;
+use rustc_lint::{EarlyContext, EarlyLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::{symbol::Ident, Span};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for redundant redefinitions of local bindings.
+    ///
+    /// ### Why is this bad?
+    /// Redundant redefinitions of local bindings do nothing and are likely to be unintended.
+    ///
+    /// ### Example
+    /// ```rust
+    /// let a = 0;
+    /// let a = a;
+    ///
+    /// fn foo(b: i32) {
+    ///    let b = b;
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let a = 0;
+    /// // no redefinition with the same name
+    ///
+    /// fn foo(b: i32) {
+    ///   // no redefinition with the same name
+    /// }
+    /// ```
+    #[clippy::version = "1.72.0"]
+    pub REDUNDANT_LOCAL,
+    correctness,
+    "redundant redefinition of a local binding"
+}
+declare_lint_pass!(RedundantLocal => [REDUNDANT_LOCAL]);
+
+impl EarlyLintPass for RedundantLocal {
+    fn check_block(&mut self, cx: &EarlyContext<'_>, block: &Block) {
+        check_block_with_bindings(cx, &mut FxHashMap::default(), block);
+    }
+
+    fn check_fn(&mut self, cx: &EarlyContext<'_>, func: FnKind<'_>, _span: Span, _node: rustc_ast::NodeId) {
+        // get the top-level block of the function's expression
+        let block = match func {
+            FnKind::Fn(_, _, _, _, _, Some(block)) => block,
+            FnKind::Closure(
+                _,
+                _,
+                Expr {
+                    kind: ExprKind::Block(block, _),
+                    ..
+                },
+            ) => block,
+            _ => return,
+        };
+
+        // collect the initial bindings from the function's parameters
+        let mut bindings = FxHashMap::default();
+        for param in &func.decl().inputs {
+            collect_introduced_bindings(&mut bindings, &param.pat);
+        }
+
+        check_block_with_bindings(cx, &mut bindings, block);
+    }
+}
```

Comment:
issue: this won't catch
```rust
fn parameter(x: i32) {
    let x = x;
    {
        let x = x;
    }
}
```
or
```rust
fn block() {
    let x = 1;
    {
        let x = x;
    }
}
```

Reply:
You should check all stmts, not just `StmtKind::Local`, that way you also include any introduced `Block`s (as currently it will only check the fn's outermost block)

Reply:
I think this can be done by using `path_res`, then using the `HirId` from that to get that node; using `cx.tcx.hir().get(hir_id)` (maybe also requiring getting the parent node). Getting the mutability is the same as you did in `collect_introduced_bindings`. This should allow you to completely omit storing the bindings, and you can combine this with checking all stmts :) Though I haven't tested this yet so, so there may be some error somewhere..

Note this will require using a late lint.